src/scores/stats/novel.py

```diff
@@ -0,0 +1,337 @@
+            all_days  float64 15.0
+        Attributes:
+            selections: {{'days123': [1, 2, 3], 'all_days': [1, 2, 3, 4]}}
+            sampling_dim: lead_day
+
+    See also:
+        `scores.stats.novel.flip_flop_index_proportion_exceeding`
+    """
+
+    if not selections:
+        result = _flip_flop_index(data, sampling_dim, is_angular=is_angular)
+    else:
+        result = xr.Dataset()
+        result.attrs["selections"] = selections
+        for key, data_subset in iter_selections(data, sampling_dim, **selections):
+            result[key] = _flip_flop_index(data_subset, sampling_dim, is_angular=is_angular)
+
+    result.attrs["sampling_dim"] = sampling_dim
+
+    return result
+
+
+def iter_selections(data: XarrayLike, sampling_dim: str, **selections: Iterable[str]):
+    """
+    Selects subsets of data along dimension sampling_dim according to
+    `selections`.
+
+    Args:
+        data (xarray.DataArray or xarray.Dataset): The data to sample from.
+        sampling_dim (str): The dimension from which to sample.
+        selections (iterable): Each supplied keyword corresponds to a
+            selection of `data` from the dimensions `sampling_dim`. The
+            key is the first element of the yielded tuple.
+
+    Yields:
+        A tuple (key, data_subset), where key is the supplied `**selections`
+        keyword, and data_subset is the `data` at the values along
+        `sampling_dim` specified by `**selections`.
+
+    Raises:
+        KeyError: values in selections are not in data[sampling_dim]
+
+    Examples:
+        >>> data = xr.DataArray(
+        ...     [0, 0.1, 0.2, 0.3, 0.4, 0.5, 0.7],
+        ...     coords={'lead_day': [1, 2, 3, 4, 5, 6, 7]}
+        ... )
+        >>> for key, data_subset in iter_selections(
+        ...         data, 'lead_day', days123=[1, 2, 3], all_days=[1, 2, 3, 4, 5, 6, 7]
+        ... ):
+        ...     print(key, ':', data_subset)
+        all_days : <xarray.DataArray (lead_day: 7)>
+        array([ 0. ,  0.1,  0.2,  0.3,  0.4,  0.5,  0.7])
+        Coordinates:
+          * lead_day  (lead_day) int64 1 2 3 4 5 6 7
+        days123 : <xarray.DataArray (lead_day: 3)>
+        array([ 0. ,  0.1,  0.2])
+        Coordinates:
+          * lead_day  (lead_day) int64 1 2 3
+
+    """
+    check_dims(data, [sampling_dim], "superset")
+
+    for key, values in selections.items():
+        try:
+            # Need copy so that attributes added in _iter_selections_with_attrs
+            # don't affect the whole dataframe but just the subset
+            data_subset = data.sel(**{sampling_dim: values}).copy(deep=False)
+        except KeyError as ex:
+            raise KeyError(
+                f"for `selections` item {str({key: values})}, not all values found in " f"dimension '{sampling_dim}'",
+            ) from ex
+
+        yield key, data_subset
+
+
+def encompassing_sector_size(data: xr.DataArray, dims: Iterable[str], skipna: bool = False):
```

Comment:
I think dims should be FlexibleDimensionTypes. Data perhaps is XarrayLike but if it needs DataArray specifically then it's fine as is.

Reply:
The `FlexibleDimensionTypes` typing alias currently includes the Optional special form which isn't true in this case. Either we can modify `FlexibleDimensionTypes` and where its used or just call `dims` type `Iterable[Hashable]`. The reason I specified `str` is because `check_dims` specifies `dims` as `list[str]` so hashable would result in an error. 

I might spend some time now refactoring these discrepancies.

Reply:
It's completely fine (indeed - better) to use a more specific type if that's what the function supports, I just wanted to check that it had been considered since in other places it tends to be optional.

Reply:
Resolving due to `Sequence[str]` being the best suited type hint for this function.